File: spark-nrt/src/main/scala/com/ebay/traffic/chocolate/sparknrt/reporting/CouchbaseClient.scala
package com.ebay.traffic.chocolate.sparknrt.reporting

import java.util.Properties
import java.util.concurrent.TimeUnit

import com.couchbase.client.java.document.JsonDocument
import com.couchbase.client.java.document.json.{JsonArray, JsonObject}
import com.couchbase.client.java.env.DefaultCouchbaseEnvironment
import com.couchbase.client.java.{Bucket, Cluster, CouchbaseCluster}
import org.slf4j.LoggerFactory

/**
  * Created by weibdai on 5/19/18.
  */
object CouchbaseClient {

  @transient private lazy val logger = LoggerFactory.getLogger(this.getClass)

  @transient private lazy val properties = {
    val properties = new Properties
    properties.load(getClass.getClassLoader.getResourceAsStream("couchbase.properties"))
    properties
  }

  // Open Couchbase bucket with settings in couchbase.properties.
  @transient lazy val cluster: Cluster = createClusterFunc()

  // Use this delegate to override the way for cluster creation.
  @transient var createClusterFunc: () => Cluster = createCluster

  private def createCluster(): Cluster = {
    try {
      val nodes = properties.getProperty("chocolate.report.couchbase.cluster")
      val user = properties.getProperty("chocolate.report.couchbase.user")
      val password = properties.getProperty("chocolate.report.couchbase.password")

      val env = DefaultCouchbaseEnvironment.builder()
        .connectTimeout(10000).queryTimeout(5000).build()
      CouchbaseCluster.create(env, nodes)
      cluster.authenticate(user, password)
    } catch {
      case e: Exception =>
        logger.error("Couchbase create cluster error.", e)
        throw e
    }
  }

  @transient lazy val bucket: Bucket = {
    try {
      val bucketName = properties.getProperty("chocolate.report.couchbase.bucket")
      cluster.openBucket(bucketName, 1200, TimeUnit.SECONDS)
    } catch {
      case e: Exception =>
        logger.error("Couchbase open bucket error.", e)
        throw e
    }
  }

  /**
    * Insert or append data into Couchbase.
    *
    * @param key     key of data
    * @param mapData value of data organized in Map[String, Any]
    */
  def upsert(key: String, mapData: Map[String, _]): Unit = {
    try {
      val jsonObject = JsonObject.empty()
      mapData.foreach(data => jsonObject.put(data._1, data._2))

      logger.debug("Couchbase upsert: " + key + " -> " + jsonObject)

      if (!bucket.exists(key)) {
        val jsonArray = JsonArray.create().add(jsonObject)
        val root = JsonObject.empty().put("data", jsonArray)
        bucket.upsert(JsonDocument.create(key, root))
      } else {
        val jsonArray = bucket.get(key).content().getArray("data")
        jsonArray.add(jsonObject)
        val root = JsonObject.empty().put("data", jsonArray)
        bucket.replace(JsonDocument.create(key, root))
      }
    } catch {
      case e: Exception => { logger.error("Couchbase upsert error.", e) }
    }
  }

  /**
    * Close bucket connection.
    */
  def close(): Unit = {
    bucket.close
    cluster.disconnect
  }
}

File: ingester/src/main/java/com/ebay/traffic/chocolate/CouchbaseClient.java
package com.ebay.traffic.chocolate;

import com.couchbase.client.java.Bucket;
import com.couchbase.client.java.Cluster;
import com.couchbase.client.java.CouchbaseCluster;
import com.couchbase.client.java.document.JsonDocument;
import com.couchbase.client.java.document.json.JsonObject;
import com.couchbase.client.java.env.CouchbaseEnvironment;
import com.couchbase.client.java.env.DefaultCouchbaseEnvironment;
import com.couchbase.client.java.query.N1qlQuery;
import com.couchbase.client.java.query.N1qlQueryResult;
import com.couchbase.client.java.query.N1qlQueryRow;
import com.google.gson.Gson;
import javafx.util.Pair;
import org.apache.commons.lang3.Validate;
import org.apache.log4j.Logger;

import java.io.FileInputStream;
import java.io.IOException;
import java.util.List;
import java.util.Properties;
import java.util.Queue;
import java.util.concurrent.LinkedBlockingDeque;
import java.util.concurrent.TimeUnit;

public class CouchbaseClient {
  private static final Logger logger = Logger.getLogger(CouchbaseClient.class);
  private static volatile CouchbaseClient INSTANCE = null;
  private static Properties cbProp;
  private final Cluster cluster;
  private Bucket bucket;
  private Queue<Pair<String, Long>> buffer;
  private Gson gson = new Gson();


  private CouchbaseClient(String configFilePath) throws IOException {
    initCBProp(configFilePath);
    CouchbaseEnvironment env = DefaultCouchbaseEnvironment.builder()
        .connectTimeout(Long.valueOf(cbProp.getProperty("seed.couchbase.connect.timeout")))
        .queryTimeout(Long.valueOf(cbProp.getProperty("seed.couchbase.query.timeout")))
        .build();
    this.cluster = CouchbaseCluster.create(env, new String[]{cbProp.getProperty("seed.couchbase.cluster")});
    try {
      this.cluster.authenticate(cbProp.getProperty("seed.couchbase.user"), cbProp.getProperty("seed.couchbase.password"));
      this.bucket = this.cluster.openBucket(cbProp.getProperty("seed.couchbase.bucket"), 1200L, TimeUnit.SECONDS);
    } catch (Exception e) {
      logger.error("Couchbase init error", e);
      throw e;
    }
    this.buffer = new LinkedBlockingDeque();
  }

  public Bucket getBucket() {
    return bucket;
  }

  private static void initCBProp(String configFilePath) throws IOException {
    cbProp = new Properties();
    cbProp.load(new FileInputStream(configFilePath + "seed.properties"));
  }

  private static void init(String configFilePath) throws IOException {
    Validate.isTrue(INSTANCE == null, "Instance should be initialized only once", new Object[0]);
    INSTANCE = new CouchbaseClient(configFilePath);
    logger.info("Initial Couchbase cluster");
  }

  public static void init(CouchbaseClient client) {
    Validate.isTrue(INSTANCE == null, "Instance should be initialized only once", new Object[0]);
    INSTANCE = client;
  }

  public static CouchbaseClient getInstance(String configFilePath) throws IOException {
    if (INSTANCE == null) {
      synchronized (CouchbaseClient.class) {
        if (INSTANCE == null) {
          init(configFilePath);
        }
      }
    }
    return INSTANCE;
  }

  public static void close() {
    if (INSTANCE == null) {
      return;
    }
    INSTANCE.bucket.close();
    INSTANCE.cluster.disconnect();
    INSTANCE = null;
  }

  public synchronized void upsert(String userId, long timestamp) {
    flushBuffer();
    try {
      UserInfo userInfo = new UserInfo();
      userInfo.setUser_id(userId);
      userInfo.setActivity_ts(Long.valueOf(timestamp));
      this.bucket.upsert(JsonDocument.create(userId, JsonObject.fromJson(this.gson.toJson(userInfo))));
      logger.debug("Adding new mapping. userId=" + userId + " timestamp=" + timestamp);
    } catch (Exception e) {
      this.buffer.add(new Pair(userId, Long.valueOf(timestamp)));
      logger.warn("Couchbase upsert operation exception", e);
    }
  }

  public List<N1qlQueryRow> query(String queryString) {
    N1qlQueryResult result = this.bucket.query(N1qlQuery.simple(queryString));
    if ((result == null) || (result.allRows() == null) || (result.allRows().size() <= 0)) {
      return null;
    }
    return result.allRows();
  }

  private void flushBuffer() {
    try {
      while (!this.buffer.isEmpty()) {
        Pair<String, Long> kv = (Pair) this.buffer.peek();
        upsert((String) kv.getKey(), ((Long) kv.getValue()).longValue());
        this.buffer.poll();
      }
    } catch (Exception e) {
      logger.warn("Couchbase upsert operation exception", e);
    }
  }
}
